src/board_builder/board_builder.py

```diff
@@ -0,0 +1,258 @@
+            if pose.target_id not in self._detector_poses_average:
+                self._detector_poses_average[pose.target_id] = PoseLocation(pose.target_id)
+            self._detector_poses_average[pose.target_id].add_matrix(np.array(pose.object_to_reference_matrix.values).reshape(4, 4), str(timestamp))
+        for label in self._detector_poses_average:
+            pose = Pose(
+                target_id=label,
+                object_to_reference_matrix=self._detector_poses_average[label].get_pose().object_to_reference_matrix,
+                solver_timestamp_utc_iso8601=str(timestamp)
+            )
+            self.detector_poses.append(pose)
+        self.pose_solver.set_detector_poses(self.detector_poses)
+
+    def collect_data(self, detector_data):
+        """ Collects data of relative position and is entered in matrix. Returns a dictionary of its corners"""
+        detector_data = self._filter_data(detector_data)
+        if not detector_data:
+            return
+        timestamp = datetime.datetime.now()
+        corners_dict = {}
+        self.target_poses = []
+        self._solve_pose(detector_data, timestamp)
+        for index, pose in enumerate(self.target_poses):
+            # R R R T
+            # R R R T
+            # R R R T
+            # 0 0 0 1
+
+            pose_values = pose.object_to_reference_matrix.values
+            pose_matrix = np.array(pose_values).reshape(4, 4)
+
+            for other_pose in self.target_poses:
+                if other_pose != pose:
+                    other_matrix_values = other_pose.object_to_reference_matrix.values
+                    other_pose_matrix = np.array(other_matrix_values).reshape(4, 4)
+                    relative_transform = self._calculate_relative_transform(pose_matrix, other_pose_matrix)
+                    matrix_index = self._find_matrix_input_index(pose.target_id, other_pose.target_id)
+
+                    if not self._relative_pose_matrix[matrix_index[0]][matrix_index[1]]:
+                        new_pose_location = PoseLocation(pose.target_id)
+                        new_pose_location.add_matrix(relative_transform, str(datetime.datetime.now()))
+                        self._relative_pose_matrix[matrix_index[0]][matrix_index[1]] = new_pose_location
+                    else:
+                        self._relative_pose_matrix[matrix_index[0]][matrix_index[1]].add_matrix(relative_transform, str(datetime.datetime.now()))
+                    self._relative_pose_matrix[matrix_index[0]][matrix_index[1]].frame_count += 1
+
+        for index, pose in enumerate(self.target_poses):
+            pose_values = pose.object_to_reference_matrix.values
+            pose_matrix = np.array(pose_values).reshape(4, 4)
+            corners_location = self._calculate_corners_location(pose_matrix, self.local_corners)
+            corners_dict[pose.target_id] = corners_location
+
+        return corners_dict
+
+    def build_board(self, detector_data):
```

Comment:
I'm not sure why you would need an input `detector_data` here. Is not the purpose of the previous `collect_data` phase to build the relative pose matrix from which we define the board points?

If we were to use a particular frame as input, wouldn't we bias our result to that input data compared to what was collected previously?

Let's plan to discuss this tomorrow.

Reply:
The section build_board comes from a slight misunderstanding I initially had of the main goal of the Board Builder task. What this specific  method is doing is that given visible markers where we know the exact pose because of the detectors (detector data), we can use the relative pose matrix to find the occluded markers. This is sort of a visual validation and completely unrelated to getting all the corner coordinates in a pre-determined reference frame (which I have yet to implement). 

For clarity, it would probably make more sense to have the public method build_board as writing the 3d corners in a file (as we discussed and which I haven't implemented yet), and perhaps call the current build_board something along the lines of find_occluded_markers_pose_given_markers_visible_to_detector. Let me know if I should do that, or if the current build_board is useful at all.

Reply:
I don't think finding occluded markers will serve much purpose other than as visual validation during development.

Reply:
My initial understanding of board builder was that the main goal was to 'figure out the location of occluded markers', and that having a board coordinate system was only one way to solve that through point registration. Since my relative pose matrix can find those occluded markers without needing a board coordinate system, I haven't even considered building one until recently. I'm guessing that a board coordinate system has its own benefits except for the fact that it can solve occluded markers because my current algorithm can do that alone.

Reply:
The board coordinate system would just be a coordinate system in which all the corner points are expressed. It can share the same origin and axes as an arbitrary marker for now.